src/ref_cnt.rs

```diff
@@ -92,7 +92,12 @@ unsafe impl<T> RefCnt for Arc<T> {
         Arc::into_raw(me) as *mut T
     }
     fn as_ptr(me: &Arc<T>) -> *mut T {
-        me as &T as *const T as *mut T
+        // SAFETY: The refcount will be incremented then decremented
+        // so we force the compiler to remove the overflow check
```

Comment:
I don't believe this is actually true/safe.

1. Ever _reaching_ unreachable_unchecked is explicitly an UB. So if the strong_count is at the max value, it simply invokes UB.
2. Even if it „just“ turned off the check inside Arc, this would not be safe. Because it would temporarily overflow to 0 and, in some other thread, could be incremented twice, decremented once, reach 0 again there and _get deallocated_.

We could use the approach without that optimization, and we could measure it that way to see if there's any actual change in speed. That would be the more elegant way and I suspect it might actually be the case.

If it turns out it is expensive, we could use a different trick:
1. Make a „hollow“ copy of an Arc, one that does not own its ref count. It could be done with eg. `std::ptr::read`.
2. Call `Arc::into_raw` on that one.
3. Turn it back into the hollow copy of Arc with `Arc::from_raw`.
4. `mem::forget` it, so we get rid of it without calling its destructor.

Reply:
That's very true - I must admit I only considered `Rc` and pasted the working code for `Arc`. I believe
```rs
fn as_ptr(me: &Arc<T>) -> *mut T {
    let ptr = Arc::into_raw(unsafe { std::ptr::read(me) } );
    ptr as *mut T
}
```
should be safe and work for both variants. Neither `ptr::read` nor `into_raw` will affect the refcount (`ptr::read` is a bitwise copy and `into_raw` calls `mem::forget` to avoid running the destructor).
The safety is much easier to verify here - we are calling `ptr::read` on a `*const T` derived from an `&T` which is guarenteed valid.

Reply:
The previous wasn't safe even for Rc, due to the fact that you could reach the unreachable_unchecked. The first point was true even for Rc.

I know that the `read` is just a bitwise copy and `into_raw` _currently_ calls the forget. In the current version, these are enough. Nevertheless, this is with the knowledge of internal implementation. The internal implementation _may_ change and these properties are not guaranteed. The only description is that `into_raw` shall be paired with `from_raw`. The `read` shall then be paired with `forget`. So it's about future proofing these for possible change in the standard library.